### PR TITLE
parse: collect external stream chunks before matching

### LINKED ISSUE #9795: nushell truncating data at 8192
### Describe the bug

A familiar error has popped up again. Nushell is truncating data at 8192 in certain conditions. I'd really like to find a smaller way to repro this but it takes a bit of data.

### Test File
[test.txt](https://github.com/nushell/nushell/files/12151402/test.txt)

### How to reproduce

1. download the test file
2. run this command on it `open \test.txt | parse -r '(?m)(?P<date>\d+/\d+/\d+)?\s+\t(?P<time>\d+:\d+:\d+)\s+(?P<user>\w+,\s\w+\s?\w+?)\t(?P<action>\w+\s?\w+)\t(?P<entry>.*)'`
3. The first 8192 is on nushell row 42. you can find this row by looking at the time column for `05:76:79`
4. The second and final error happens 8192 bytes later at nushell row 86 at `07:13:14`

### Expected behavior

don't truncate

### Screenshots

row 42 problem
image

row 86 problem
image

### Configuration

|key|value|
|-|-|
|version|0.82.1|
|branch|main|
|commit_hash|9db0d6bd34a99805c6da296688aa186778be5a86|
|build_os|windows-x86_64|
|build_target|x86_64-pc-windows-msvc|
|rust_version|rustc 1.69.0 (84c898d65 2023-04-16)|
|rust_channel|1.69.0-x86_64-pc-windows-msvc|
|cargo_version|cargo 1.69.0 (6e9a83356 2023-04-12)|
|build_time|2023-07-24 13:22:15 -05:00|
|build_rust_channel|release|
|allocator|mimalloc|
|features|dataframe, default, extra, sqlite, trash, which, zip|
|installed_plugins|custom-value generate, custom-value generate2, custom-value update, from eml, from ics, from ini, from parquet, from vcf, gstat, highlight, hist, inc, json path, nu-example-1, nu-example-2, nu-example-3, plot, xyplot|

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by panicbit**
# Description
This PR implements the workaround discussed in #9795, i.e. having `parse` collect an external stream before operating on it with a regex.

- Should close #9795 

# User-Facing Changes

- `parse` will give the correct output for external streams
- increased memory and time overhead due to collecting the entire stream (no short-circuiting)

# Tests + Formatting

- no tests that weren't already broken fail
- formatting is checked
- added test case

# After Submitting

### CODE REVIEW

**Comment by fdncred** [comment]
nice work @panicbit. I've tested it out and it seems to work. nice test too!

File: crates/nu-command/tests/commands/parse.rs [hunk 1]
```diff
@@ -211,4 +211,19 @@ mod regex {
...
+                cwd: dirs.test(), pipeline(
+                r#"
+                    "abcdefghijklmnopqrstuvwxyz" * 1000 | save --force data.txt;
+                    open data.txt | parse --regex "(abcdefghijklmnopqrstuvwxyz)" | length
+                "#
+            ));
+
+            assert_eq!(actual.out, "1000");
+        })
+    }
```
Comment (amtoine) [inline crates/nu-command/tests/commands/parse.rs:228]: maybe we can have a simpler test like this?
```suggestion
    #[test]
    fn parse_handles_external_stream_chunking() {
        let actual = nu!(
            pipeline(
            r#"
                "abcdefghijklmnopqrstuvwxyz" * 1000 | save --force data.txt;
                open data.txt | parse --regex "(abcdefghijklmnopqrstuvwxyz)" | length
            "#
        ));

        assert_eq!(actual.out, "1000");
    }
```

not sure if `save` requires to be in a playground, though, i might be wrong on that one :wink:

Comment (panicbit) [reply crates/nu-command/tests/commands/parse.rs:228]: The CWD of the test and `nu!` macro seems to be the source folder, so the Playground (or something else that gives me a reasonable dir to write to) seems to be needed.

Comment (fdncred) [reply crates/nu-command/tests/commands/parse.rs:228]: Thanks @panicbit for investigating.

**Comment by fdncred** [comment]
We (nearly) always want to stream, but until we can figure out a way to move forward with streaming, this seems to be the best approach. Thanks for figuring it out @panicbit!

**Comment by fdncred** [comment]
Just added `release-note-mention` because there have a been a couple PRs for `parse` which makes it more reliable.

File: crates/nu-command/tests/commands/parse.rs (hunk 1 above)
Comment (amtoine) [reply crates/nu-command/tests/commands/parse.rs:228]: @panicbit 
that's interesting to know, thanks :pray: 

TL;DR: when writing a file to disk in the tests, a playground is needed :ok_hand:
